## Importing data

In [48]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [49]:
df=pd.read_csv("Telecom_churn.csv")
df.head()

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                 no   
1    OH             107        415     371-7191                 no   
2    NJ             137        415     358-1921                 no   
3    OH              84        408     375-9999                yes   
4    OK              75        415     330-6626                yes   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   
2             41.38  ...              110             10.30   
3             50.90  ...               88              5.26   
4             28.34  ...              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False  

[5 rows x 21 columns]

In [50]:
df.isnull().sum()

state                     0
account length            0
area code                 0
phone number              0
international plan        0
voice mail plan           0
number vmail messages     0
total day minutes         0
total day calls           0
total day charge          0
total eve minutes         0
total eve calls           0
total eve charge          0
total night minutes       0
total night calls         0
total night charge        0
total intl minutes        0
total intl calls          0
total intl charge         0
customer service calls    0
churn                     0
dtype: int64

**Dataset contains no missing values**

## Data preparation

In [51]:
# Converting yes/no to 0/1:
df['international plan']=df['international plan'].apply(lambda x: 1 if x=='yes' else 0)
df['voice mail plan']=df['voice mail plan'].apply(lambda x: 1 if x=='yes' else 0)
df['churn']=df['churn'].apply(lambda x: 1 if x==True else 0)
df.head()

state  account length  area code phone number  international plan  \
0    KS             128        415     382-4657                   0   
1    OH             107        415     371-7191                   0   
2    NJ             137        415     358-1921                   0   
3    OH              84        408     375-9999                   1   
4    OK              75        415     330-6626                   1   

   voice mail plan  number vmail messages  total day minutes  total day calls  \
0                1                     25              265.1              110   
1                1                     26              161.6              123   
2                0                      0              243.4              114   
3                0                      0              299.4               71   
4                0                      0              166.7              113   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   
2             41.38  ...              110             10.30   
3             50.90  ...               88              5.26   
4             28.34  ...              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0  

[5 rows x 21 columns]

In [52]:
# Creating dummies for 'state' variable:
df=pd.merge(df,
            pd.get_dummies(df.state,drop_first=True),
            left_index=True,
            right_index=True,
            how='inner')
df.head()

state  account length  area code phone number  international plan  \
0    KS             128        415     382-4657                   0   
1    OH             107        415     371-7191                   0   
2    NJ             137        415     358-1921                   0   
3    OH              84        408     375-9999                   1   
4    OK              75        415     330-6626                   1   

   voice mail plan  number vmail messages  total day minutes  total day calls  \
0                1                     25              265.1              110   
1                1                     26              161.6              123   
2                0                      0              243.4              114   
3                0                      0              299.4               71   
4                0                      0              166.7              113   

   total day charge  ...  SD  TN  TX  UT  VA  VT  WA  WI  WV  WY  
0             45.07  ...   0   0   0   0   0   0   0   0   0   0  
1             27.47  ...   0   0   0   0   0   0   0   0   0   0  
2             41.38  ...   0   0   0   0   0   0   0   0   0   0  
3             50.90  ...   0   0   0   0   0   0   0   0   0   0  
4             28.34  ...   0   0   0   0   0   0   0   0   0   0  

[5 rows x 71 columns]

In [53]:
# Dropping not needed columns
df=df.drop(['state','account length','area code','phone number'],axis=1)
df.head()

international plan  voice mail plan  number vmail messages  \
0                   0                1                     25   
1                   0                1                     26   
2                   0                0                      0   
3                   1                0                      0   
4                   1                0                      0   

   total day minutes  total day calls  total day charge  total eve minutes  \
0              265.1              110             45.07              197.4   
1              161.6              123             27.47              195.5   
2              243.4              114             41.38              121.2   
3              299.4               71             50.90               61.9   
4              166.7              113             28.34              148.3   

   total eve calls  total eve charge  total night minutes  ...  SD  TN  TX  \
0               99             16.78                244.7  ...   0   0   0   
1              103             16.62                254.4  ...   0   0   0   
2              110             10.30                162.6  ...   0   0   0   
3               88              5.26                196.9  ...   0   0   0   
4              122             12.61                186.9  ...   0   0   0   

   UT  VA  VT  WA  WI  WV  WY  
0   0   0   0   0   0   0   0  
1   0   0   0   0   0   0   0  
2   0   0   0   0   0   0   0  
3   0   0   0   0   0   0   0  
4   0   0   0   0   0   0   0  

[5 rows x 67 columns]

## Fitting logistic regression

In [98]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression

In [99]:
# Randomly splitting dataset on train and test parts:
x=df.drop('churn',axis=1)
y=df['churn']
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=True)

In [100]:
# Scaling data to unit variance
std=StandardScaler()
# Fitting and transforming x_train
x_train_scaled=std.fit_transform(x_train)
# Transforming x_test using already fitted StandardScaler
x_test_scaled=std.transform(x_test)

In [101]:
# Creating GridSearchCV instance to find best parameters using k-fold method
params={'C':[0.001,0.01,1,100,1000]}
gridCV=GridSearchCV(LogisticRegression(),params,cv=5)

In [102]:
gridCV.fit(x_train_scaled,y_train)

GridSearchCV(cv=5, error_score=nan,
             estimator=LogisticRegression(C=1.0, class_weight=None, dual=False,
                                          fit_intercept=True,
                                          intercept_scaling=1, l1_ratio=None,
                                          max_iter=100, multi_class='auto',
                                          n_jobs=None, penalty='l2',
                                          random_state=None, solver='lbfgs',
                                          tol=0.0001, verbose=0,
                                          warm_start=False),
             iid='deprecated', n_jobs=None,
             param_grid={'C': [0.001, 0.01, 1, 100, 1000]},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=0)

In [103]:
print("Best parameter: ",gridCV.best_params_)
print("Accuracy: ",gridCV.cv_results_['mean_test_score'][1])

Best parameter:  {'C': 0.01}
Accuracy:  0.8638411647729269


## Checking model on test dataset:

In [104]:
# Calculating predictions
predictions=gridCV.best_estimator_.predict(x_test_scaled)

In [116]:
from sklearn.metrics import precision_recall_fscore_support

In [121]:
print('Accuracy:', (predictions==y_test).mean())
print("Precision: ",precision_recall_fscore_support(y_test,predictions)[0][0])
print("Recall: ",precision_recall_fscore_support(y_test,predictions)[1][0])

Accuracy: 0.8515742128935532
Precision:  0.8604294478527608
Recall:  0.9859402460456942


Accuracy of the model is 85%, which is quite high and consistent with the accuracy obtained during training, so we can say that during training there weren't overfitting and underfitting. 

Precision is 86%, which means that out of all customers labeled as churn ones 86% of them really churned.

Recall is 98%, which shows that out of all customers that churned our model will detect 98% of them